File: xcd.pl
#!/usr/bin/env perl

use strict;
use warnings;

my $db;
my %hash;

#
# read database
#
sub ReadDatabase {
	delete @hash{keys %hash};

	open MYFILE, $db or die $!;
	my $key;
	while (my $line = <MYFILE>) {
		chomp $line;
		if ($line =~ /^([^:]+):([^:]+)$/) {
			my ($k, $v) = ($1, $2);
			$hash{$k} = $v;
		}
	}
	close MYFILE;
	# print map { "$_:$hash{$_}\n" } keys %hash;
}

#
# lookup table
#
sub LookupPath {
	my $key = shift;
	return $hash{$key} if exists $hash{$key};
	return '';
}

#
# locate database
#
$db = `dirname $0`;
chomp $db;
$db .= '/xcd.db';
# print "$db";

#
# parse arguments
#
my $script = '';
if (0 == @ARGV) {
	$script .= "cd -";
# } elsif (1 == @ARGV) {
} elsif (2 == @ARGV) {
	my $prj = $ARGV[1];
	my $pre = "~/my/app/$ARGV[0]";
	my $pos = "$prj/assets/www";
	my $rz = "$pre/$pos";
	$script .= "cd $rz;title $prj\n";
} elsif ($ARGV[0] =~ /^([^:]+):([^:]+)$/) {
	my ($site, $path) = ($1, $2);
	if ($path =~ m/^~/) {
		# my $home = $ENV{HOME};
		# $path =~ s/~/$home/;
		$path =~ s/~//;
	}
	$script .= "cd ~/site/$site/$path" . "\n";
} else {
	ReadDatabase($db);
	my $rz = LookupPath($ARGV[0]);
	if (not $rz) {
		print "define '$ARGV[0]' : ";
		my $path = <STDIN>;
		open (MYFILE, ">>$db");
		print MYFILE "$ARGV[0]:$path";
		close (MYFILE);

		ReadDatabase($db);
		$rz = LookupPath($ARGV[0]);
		if (not $rz) { exit 1; }
	}
	$script .= "mkdir -p $rz;cd $rz";
}

#
# generate /tmp/x
#
open (MYFILE, ">/tmp/x");
print MYFILE "$script";
close (MYFILE);
exit 0;

File: xcd.py
#!/usr/bin/env python3


class Structure:
    def __init__(self, **entries):
        self.__dict__.update(entries)


# helpers
import pprint
pp = pprint.PrettyPrinter(indent=4)

# get script path
import os
py_path = os.path.dirname(os.path.realpath(__file__))

# read config
import json
cfg_file = os.path.join(py_path, 'config.json')
with open(cfg_file) as fh:
    jsn = json.load(fh)
    cfg = Structure(**jsn)

# determin key to lookup
import sys
if len(sys.argv) < 2:
    print('nowhere to go, abort!')
    exit(1)
key = sys.argv[1]
# print('key', key)

# convert key to path
path = key
import os
if '.' == key[0]:
    path = os.path.abspath(key)
else:
    # load database
    dbname = sys.argv[0][:-2] + 'db'
    lst = dbname.split('/')
    lst.insert(-1, 'data/{}'.format(cfg.context))
    dbname = '/'.join(lst)
    # print('dbname', dbname)
    dbfh = open(dbname, 'r+')
    db = {}
    for line in dbfh:
        (k, v) = line.strip().split(':')
        db[k] = v
    # pp.pprint(db)

    # lookup now
    if key in db:
        # found
        path = db[key]
    else:
        # new
        path = input("define '{}' : ".format(key))
        dbfh.write("{}:{}\n".format(key, path))
    dbfh.close()
# print('path', path)

# generate shell script
from string import Template
title = '/'.join(path.split('/')[-2:])
d = dict(path=path, title=title)
template = Template('mkdir -p $path;cd $path;title $title;[ -e .xcd.rc ] && . .xcd.rc')
script = template.substitute(d)
# print('script', script)

# done
print(script)
